### PR TITLE
fix scaler race condition via pending placeholder pod detection

### PR DESCRIPTION
**Pull request by shaneknapp**
This resolves https://github.com/berkeley-dsep-infra/jupyterhub-k8s-node-placeholder/issues/25

## Summary

Three race conditions fixed in the node-placeholder scaler, plus four crash bugs found during a test coverage audit and log review.

## Race condition 1: placeholder pod eviction

When a user logs in, their pod can land on the node running the placeholder. The placeholder gets evicted, but the node's memory usage stays low while the image pulls. On the next scaler cycle, the node looks free (no placeholder, low utilization), so the scaler decrements the replica count before the placeholder reschedules. The pool is left with no warm spare.

The fix checks whether any placeholder pod for the pool is already Pending before applying the reduction. If one is pending, the reduction is suppressed for that cycle.

New functions: `compute_replica_count()` (pure, unit-testable) and `any_placeholder_pod_pending()` (per-pool filtering via `node_selector`).

## Race condition 2: node oscillation after scale-to-zero

When GKE's autoscaler cordons and drains all nodes in a pool (after a calendar-driven scale-to-zero, for example), the scaler correctly sets replicas=1 once no schedulable nodes remain. GKE then provisions new nodes for the pending placeholder. With a BALANCED location policy, GKE provisions one node per zone rather than one total. Those extra nodes look free to the scaler, which immediately counts them toward its reduction and drives replicas back to 0 before the placeholder stabilizes. The result is an indefinite cycle: cordon, scaler sets 1, GKE provisions 2, scaler reduces to 0, GKE drains, repeat.

The fix adds a per-node grace period (default 300s, tunable via `--node-grace-period`). The scaler records the first time it observes each node using `time.perf_counter()` and skips nodes younger than the grace period when computing the reduction count. Nodes that are unschedulable or already running a placeholder pod are not affected. First-seen timestamps are cleaned up when nodes leave the cluster. If the scaler pod restarts, the timestamps reset and all nodes are treated as new.

I have also set the autoscaler location policy to ANY.

First-seen tracking is in `update_node_first_seen()` so it can be tested independently without mocking the full `main()` loop.

## Race condition 3: recently-freed node triggers premature reduction

When a user logs in and their pod lands on the placeholder's node, the placeholder is evicted and rescheduled to a freshly provisioned node. If the user logs out quickly, the original node drops below the memory threshold. The scaler counts it as free and reduces replicas to zero, even though the placeholder is running fine on another node. The placeholder terminates, both nodes drain, GKE removes them, and the next user waits for a cold start.

Observed in production on June 12, 2026: a single brief login produced six placeholder pod replacement cycles and significant node churn in an otherwise idle pool.  This was confirmed via Prometheus pod phase metrics and GKE autoscaler event logs

The fix tracks when each node was last above the utilization threshold or hosting a placeholder pod, using a new `node_last_above_threshold` dict that is updated every cycle. When a node drops below threshold or loses its placeholder, it is excluded from reduction for `--node-grace-period` seconds (default 300s, same argument as race condition 2).

Tracking is in `update_node_last_above_threshold()` for independent unit testing.

## Crash bugs in `get_usable_resources`

Three bugs found during a coverage audit of the existing tests (originally written with Copilot):

1. `AttributeError` when a pool has no pods: `requested_resources.get(pool)` returned `None`, and the chained `.get(node)` call raised. This hits any freshly provisioned pool.
2. `TypeError` when a node has no pods but its pool does: the pool existed in `requested_resources` but the node did not, so `.get(node)` returned `None`.
3. `ZeroDivisionError` when allocatable CPU or memory is 0: `get_allocatable_resources_by_pool` already defaults to 0 on parse failure, so this path was reachable.

Fix: replace `.get(pool).get(node)` with `.get(pool, {}).get(node, {"cpu_m": 0, "mem_mi": 0})` and guard both ratio divisions against zero denominators.

## Crash bug in `calendar_parser.get_events`

`get_events()` strips HTML from event descriptions by calling `re.sub()` on `ev.description`. Some calendar events have no description field, so `ev.description` is `None`. Passing `None` to `re.sub()` raises a `TypeError` and kills the scaler pod. The crash shows up in `kubectl logs --previous` on the running pod.

The fix skips HTML stripping when `ev.description` is `None`.

## Calendar count of zero was indistinguishable from no calendar event

The scaler read the calendar override with `replica_count_overrides.get(pool_name, 0)` and then gated on `if calendar_replica_count > 0`. Those two lines collapse two different situations into the same value. A pool with no calendar event at all got `0`, and a pool with an event that deliberately sets the count to `0` also got `0`. Neither could pass a `> 0` gate, so an explicit zero from the calendar never took the calendar branch. It fell through to the node-reduction path instead.

This has not caused a visible incident, because the fallthrough path happened to arrive at zero anyway in the cases we were running. The behavior was right by coincidence, not by construction, and it was one config change away from being wrong.

It matters more now that Berkeley treats the calendar as the canonical source for replica counts rather than the config. An event that sets a pool to zero is a deliberate instruction and needs to beat the config value the same way any other count does. cal-icor has the concrete version of this: an evening cool-off window that sets `base: 0` from 7pm to 7am. That is a real instruction to run no warm placeholder overnight, and it should not have to depend on the reduction logic reaching the same answer on its own.

The fix separates presence from value. The default is now `None`:

```python
calendar_replica_count = replica_count_overrides.get(pool_name, None)
```

and both the branch in `compute_replica_count` and the matching log line in `_process_pool` gate on `is not None`. `None` means no event covers this pool right now. Any integer, zero included, is something the calendar actually said, and it wins when `calendar_override_enabled` is set.

`>= 0` would also have separated the two cases, but only as long as nothing negative can reach the gate. The calendar branch returns its value directly with no `max(..., 0)` floor, so a typo like `base: -1` in an event description would have been passed straight through as a replica count. Rather than defend against that at the gate, `get_replica_counts` now rejects it at the point of parsing:

```python
if count < 0:
    log.error(f"Count {count} for pool {pool_name} is negative; skipping.")
    continue
```

The overrides dict is therefore guaranteed to hold non-negative integers, which is what lets the sentinel be about presence instead of sign.

The logs read better as a side effect. `Calendar replica count for pool base: None` says plainly that no event applies. A sentinel of `-1` would have printed a number that looks like a count and means the opposite.

Tests updated for the new contract: `get_replica_counts` now has coverage for a negative count being skipped and for an explicit `0` being stored rather than discarded, and the `compute_replica_count` cases that previously passed `0` to mean "no event" pass `None`.

## Logging fixes

- Restored the calendar override log message that was silently dropped in the original control flow.
- Converted an ambiguous `if/if/return` block to `if/elif/else` so the log message matches the branch actually taken.
- Split a misleading "placeholder pod is running OR node is unschedulable" log into two separate messages.

## Tests

126 total (up from 98). New tests cover: pending placeholder detection, crash paths in `get_usable_resources`, calendar override edge cases (an explicit count of 0 is honored, negative counts are rejected), `compute_replica_count` priority logic, `update_node_first_seen`, `update_node_last_above_threshold`, and `get_events` with a no-description event.

### CODE REVIEW

File: helm/node-placeholder/templates/deployment.yaml [hunk 1]
```diff
@@ -44,6 +44,7 @@ spec:
             - --cpu-threshold={{ .Values.cpuThreshold }}
             - --memory-threshold={{ .Values.memoryThreshold }}
             - --strategy={{ .Values.scalingStrategy | default "balanced" }}
+            - --node-grace-period={{ .Values.nodeGracePeriod }}
```
Comment (yijunge-ucb) [inline helm/node-placeholder/templates/deployment.yaml]: - --node-grace-period={{ .Values.nodeGracePeriod  |  default 600 }}
If .Values.nodeGracePeriod ever renders empty, the container will crash.

File: node-placeholder-scaler/scaler/scaler.py [hunk 2]
```diff
@@ -179,27 +172,66 @@ def placeholder_pod_running_on_node(node_name, namespace, label_selector):
...
+        return False
+
+    except client.exceptions.ApiException as e:
+        log.error(f"Kubernetes API error: {e}")
+        return False
+
+
+def compute_replica_count(
+    modified_replica,
+    config_replica_count,
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/scaler/scaler.py]: replace with "override_replica_count"

File: node-placeholder-scaler/scaler/scaler.py [hunk 3]
```diff
@@ -179,27 +172,66 @@ def placeholder_pod_running_on_node(node_name, namespace, label_selector):
...
+    calendar_replica_count,
+    calendar_override_enabled,
+    has_pending_placeholder=False,
+):
+    """Return the target placeholder replica count for a pool.
+
+    Calendar override takes priority. If a placeholder pod is Pending (evicted
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/scaler/scaler.py:214]: "Calendar override takes priority. If a placeholder pod is Pending (evicted
but not yet rescheduled), suppress reduction and fall back to
override_replica_count,  the same pre-reduction baseline used to compute
modified_replica (replica_count_overrides.get(pool_name, pool_config["replicas"])), so the deployment isn't scaled down during the window before the pod finds a new home, and any calendar-provided
count is preserved even when calendar override isn't enabled." 

This is actually very important for Berkeley deployments because now we prefer adding calendar events over changing configs.

File: node-placeholder-scaler/scaler/scaler.py [hunk 4]
```diff
@@ -179,27 +172,66 @@ def placeholder_pod_running_on_node(node_name, namespace, label_selector):
...
+    but not yet rescheduled), suppress reduction so the deployment isn't scaled
+    down during the window before the pod finds a new home.
+    """
+    if calendar_replica_count > 0 and calendar_override_enabled:
+        return calendar_replica_count
+    elif has_pending_placeholder:
+        return config_replica_count
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/scaler/scaler.py]: line 221 : return max(override_replica_count, 0)

File: node-placeholder-scaler/scaler/scaler.py [hunk 5]
```diff
@@ -214,45 +246,200 @@ def make_deployment(pool_name, template, node_selector, resources, replicas):
...
+                )
+            else:
+                log.info(
+                    f"Node {node} has sufficient resources (Strategy: {strategy}, CPU free ratio: {cpu_free_ratio:.2f}, Memory free ratio: {mem_free_ratio:.2f})."
+                )
+                node_placeholder_deployment_reduction += 1
+
+    calendar_replica_count = replica_count_overrides.get(pool_name, 0)
+    config_replica_count = pool_config["replicas"]
+    modified_replica = (
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/scaler/scaler.py]: override_replica_count = replica_count_overrides.get(
      pool_name, pool_config["replicas"]
     )
modified_replica = override_replica_count - node_placeholder_deployment_reduction

File: node-placeholder-scaler/scaler/scaler.py [hunk 6]
```diff
@@ -214,45 +246,200 @@ def make_deployment(pool_name, template, node_selector, resources, replicas):
...
+            f"Suppressing reduction for pool {pool_name}: placeholder pod is Pending."
+        )
+    else:
+        log.info(
+            f"Reducing {pool_name} placeholder deployment replicas by {node_placeholder_deployment_reduction} based on node resources."
+        )
+
+    replica_count = compute_replica_count(
+        modified_replica,
+        config_replica_count,
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/scaler/scaler.py]: line 416: override_replica_count,

File: node-placeholder-scaler/tests/test_scaler.py [hunk 7]
```diff
@@ -690,3 +784,260 @@ def test_node_name_passed_to_api(self, mock_api_cls, mock_incluster, mock_kube):
...
+    def test_normal_with_reduction(self):
+        """When a node has spare capacity, reduction brings count to 0."""
+        assert compute_replica_count(0, 1, 0, False) == 0
+
+    def test_pending_suppresses_reduction(self):
+        """Race condition: placeholder evicted but not yet rescheduled."""
+        assert compute_replica_count(0, 1, 0, False, has_pending_placeholder=True) == 1
+
+    def test_pending_returns_config_not_modified(self):
+        """Floor is config_replica_count, not modified_replica, when pending."""
```
Comment (yijunge-ucb) [inline node-placeholder-scaler/tests/test_scaler.py]: """Floor is override_replica_count, not modified_replica, when pending."""
